### PR TITLE
[SPARK-42249][SQL] Refining html link for documentation in error messages.

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to refine html link for documentation in error messages by introducing `SPARK_DOC_ROOT` into `core/src/main/scala/org/apache/spark/package.scala` that contains global directory for documentation root link: `https://spark.apache.org/docs/latest`

### Why are the changes needed?

To improve error class readability and make sure using right root document directory across all source codes.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Updated the whole existing tests related to this changes.

Basically, run `./build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite*"`

### CODE REVIEW

**Comment by itholic** [comment]
cc @srielau @MaxGekk This basically refines the doc html string in error class.

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -496,7 +496,7 @@
   },
   "ENCODER_NOT_FOUND" : {
     "message" : [
-      "Not found an encoder of the type <typeName> to Spark SQL internal representation. Consider to change the input type to one of supported at https://spark.apache.org/docs/latest/sql-ref-datatypes.html."
+      "Not found an encoder of the type <typeName> to Spark SQL internal representation. Consider to change the input type to one of supported at '<docroot>/sql-ref-datatypes.html'."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:499]: Could you elaborate why do you need quoting by '' around the link.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -104,6 +104,9 @@ private[spark] object Utils extends Logging {
   val MAX_DIR_CREATION_ATTEMPTS: Int = 10
   @volatile private var localRootDirs: Array[String] = null
 
+  /** Root directory for Apache Spark documentation. */
+  val DOC_ROOT_DIR = "https://spark.apache.org/docs/latest"
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:108]: Have you considered other places for the value? For instance, shouldn't it be placed in SparkBuildInfo:
https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/package.scala#L98

@HyukjinKwon @dongjoon-hyun @srowen @cloud-fan WDYT?

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:108]: +1

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:108]: +1, too

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:108]: +1

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (itholic) [reply core/src/main/resources/error/error-classes.json:499]: I just mimic the behavior of the existing error messages.

e.g.
```json
  "WRONG_NUM_ARGS" : {
    "message" : [
      "The <functionName> requires <expectedNum> parameters but the actual number is <actualNum>."
    ],
    "subClass" : {
      "WITHOUT_SUGGESTION" : {
        "message" : [
          "Please, refer to 'https://spark.apache.org/docs/latest/sql-ref-functions.html' for a fix."
        ]
      },
```

Do we want to just remove the '' ??

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (itholic) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Updated. Thanks!! :-)

**Comment by MaxGekk** [comment]
@itholic Rebasing on the recent master is not enough, could you fix the compiler error:
```
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:141:63: not enough arguments for method cannotFindEncoderForTypeError: (typeName: String, docroot: String)org.apache.spark.SparkUnsupportedOperationException.
[error] Unspecified value parameter docroot.
[error]       throw QueryExecutionErrors.cannotFindEncoderForTypeError(t.toString)
[error]                                                               ^
[error] one error found
```

**Comment by itholic** [comment]
Oh... yeah just fixed the test.
Thanks for catching out!

**Comment by MaxGekk** [comment]
> by introducing Utils.DOC_ROOT_DIR 

@itholic Could you update PR's description, please.

**Review by MaxGekk — APPROVED** [review]

**Comment by itholic** [comment]
Updated. Thanks!

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.4.
Thank you, @itholic and @dongjoon-hyun @HyukjinKwon @cloud-fan for review.
